pvlib/modelchain.py

```diff
@@ -1005,77 +1152,88 @@ def complete_irradiance(self, weather):
+
+        return self
 
-        icolumns = set(self.weather.columns)
+    def _complete_irradiance(self, weather):
+        icolumns = set(weather.columns)
         wrn_txt = ("This function is not safe at the moment.\n" +
                    "Results can be too high or negative.\n" +
                    "Help to improve this function on github:\n" +
                    "https://github.com/pvlib/pvlib-python \n")
 
         if {'ghi', 'dhi'} <= icolumns and 'dni' not in icolumns:
             clearsky = self.location.get_clearsky(
-                self.weather.index, solar_position=self.solar_position)
-            self.weather.loc[:, 'dni'] = pvlib.irradiance.dni(
-                self.weather.loc[:, 'ghi'], self.weather.loc[:, 'dhi'],
-                self.solar_position.zenith,
+                weather.index, solar_position=self.results.solar_position)
+            weather.loc[:, 'dni'] = pvlib.irradiance.dni(
+                weather.loc[:, 'ghi'], weather.loc[:, 'dhi'],
+                self.results.solar_position.zenith,
                 clearsky_dni=clearsky['dni'],
                 clearsky_tolerance=1.1)
         elif {'dni', 'dhi'} <= icolumns and 'ghi' not in icolumns:
             warnings.warn(wrn_txt, UserWarning)
-            self.weather.loc[:, 'ghi'] = (
-                self.weather.dni * tools.cosd(self.solar_position.zenith) +
-                self.weather.dhi)
+            weather.loc[:, 'ghi'] = (
+                weather.dhi + weather.dni *
+                tools.cosd(self.results.solar_position.zenith)
+            )
         elif {'dni', 'ghi'} <= icolumns and 'dhi' not in icolumns:
             warnings.warn(wrn_txt, UserWarning)
-            self.weather.loc[:, 'dhi'] = (
-                self.weather.ghi - self.weather.dni *
-                tools.cosd(self.solar_position.zenith))
-
-        return self
+            weather.loc[:, 'dhi'] = (
+                weather.ghi - weather.dni *
+                tools.cosd(self.results.solar_position.zenith))
 
     def _prep_inputs_solar_pos(self, kwargs={}):
         """
         Assign solar position
         """
-        self.solar_position = self.location.get_solarposition(
-            self.weather.index, method=self.solar_position_method,
+        self.results.solar_position = self.location.get_solarposition(
+            self.times, method=self.solar_position_method,
             **kwargs)
         return self
 
     def _prep_inputs_airmass(self):
         """
         Assign airmass
         """
-        self.airmass = self.location.get_airmass(
-            solar_position=self.solar_position, model=self.airmass_model)
+        self.results.airmass = self.location.get_airmass(
+            solar_position=self.results.solar_position,
+            model=self.airmass_model)
         return self
 
     def _prep_inputs_tracking(self):
         """
         Calculate tracker position and AOI
         """
         self.tracking = self.system.singleaxis(
```

Comment:
Add `tracking` attribute to `ModelChainResult` and deprecated `_deprecated_attrs`.

Reply:
Maybe we should replace the hard-coded `_deprecated_attrs` list with

```python
_deprecated_attrs = tuple(field.name for field in dataclasses.fields(ModelChainResults))
```

(https://docs.python.org/3/library/dataclasses.html#dataclasses.fields) so we don't have to maintain the list (already forgot to deprecate `ac`).

Reply:
No objection, though I guess I like the more explicit list of deprecated attributes and it's not something we're going to maintain for very long.